### ISSUE TITLE
Uploading of profile picture doesn't work on IE8

### ISSUE THREAD

**Issue by SergioBertolinSG** [owncloud/core#18007]
### Steps to reproduce

Using internet explorer 8 do:
1. Go to personal page.
2. Upload a picture (anyone of any size).

### Expected behaviour
Profile picture is updated.

### Actual behaviour
Profile picture is not shown.

### Server configuration

Operating system:
Ubuntu 14.04 LTS

Web server:
Apache

Database:
MySQL

PHP version:
5.5.9

ownCloud version: (see ownCloud admin page)
{"installed":true,"maintenance":false,"version":"8.1.1.1","versionstring":"8.1.1 beta1","edition":"Enterprise"}

Updated from an older ownCloud or fresh install:
fresh

List of activated apps:
Default ones.

Are you using external storage, if yes which one: local/smb/sftp/...
No.

Are you using encryption: 
No.

### Client configuration
Internet Explorer 8

**Comment by karlitschek** [comment]
@DeepDiver1975

**Comment by DeepDiver1975** [comment]
@SergioBertolinSG any errors in the error console?

**Comment by DeepDiver1975** [comment]
maybe this is simply broken because of this https://github.com/owncloud/core/issues/18011

**Comment by PVince81** [comment]
Getting 400 Bad request in IE8 on stable8.1 (after https://github.com/owncloud/core/issues/18011 was fixed)

**Comment by PVince81** [comment]
No errors in owncloud.log, error_log nor JS console on upload.

**Comment by PVince81** [comment]
If only we had a single code path for uploading files in JS... it would help not having such corner cases. Regular upload works.

I'll try and debug this.

**Comment by PVince81** [comment]
It goes to "No image or data provided" in the code.

**Comment by PVince81** [comment]
The app framework request doesn't seem to contain any files in the "files" attribute, see
```
$this                            = (object|OC\Core\Avatar\AvatarController[8]);
  $this->avatarManager           = (object|OC\AvatarManager[0]);
  $this->cache                   = (object|OC\Cache\File[1])+;
  $this->l                       = (object|OC_L10N[8])+;
  $this->userManager             = (object|OC\User\Manager[4])+;
  $this->userSession             = (object|OC\User\Session[3])+;
  $this->appName                 = (string[4]) 'core';
  $this->request                 = (object|OC\AppFramework\Http\Request[7]);
    $this->request->inputStream  = (string[11]) 'php://input';
    $this->request->content      = (null);
    $this->request->items        = (array[11]);
      $this->request->items['params'] = (array[0]);
      $this->request->items['get'] = (array[0]);
      $this->request->items['post'] = (array[0]);
      $this->request->items['files'] = (array[0]);
      $this->request->items['server'] = (array[40])+;
      $this->request->items['env'] = (array[0]);
      $this->request->items['cookies'] = (array[8])+;
      $this->request->items['urlParams'] = (array[1]);
        $this->request->items['urlParams']['_route'] = (string[22]) 'core.avatar.postAvatar';
      $this->request->items['parameters'] = (array[1]);
        $this->request->items['parameters']['_route'] = (string[22]) 'core.avatar.postAvatar';
      $this->request->items['method'] = (string[4]) 'POST';
      $this->request->items['requesttoken'] = (string[30]) 'Mp6TJscBsO5zXU7TI+ErHpj4+xsJS8';
    $this->request->allowedKeys  = (array[10])+;
    $this->request->secureRandom = (object|OC\Security\SecureRandom[2])+;
    $this->request->config       = (object|OC\AllConfig[3])+;
    $this->request->requestId    = (string[0]);
  $this->*OCP\AppFramework\Controller*responders = (array[1])+;
```

**Comment by PVince81** [comment]
When the request is done from Chrome, it looks like this:#
```
      $this->request->items['files'] = (array[1]);
        $this->request->items['files']['files'] = (array[5]);
          $this->request->items['files']['files']['name'] = (array[1])+;
          $this->request->items['files']['files']['type'] = (array[1])+;
          $this->request->items['files']['files']['tmp_name'] = (array[1])+;
          $this->request->items['files']['files']['error'] = (array[1])+;
          $this->request->items['files']['files']['size'] = (array[1])+;
```

So either IE8 is not sending the file properly or the App Framework's request parsing is missing something for IE8 files.

Please note that uploading files in the web UI is going through ajax/upload.php which doesn't use the app framework, which is why it work there.

**Comment by PVince81** [comment]
Another quickfix would be to remove the "Upload file" button in IE8...
People can still first upload their files in the files app and then use "Select from existing file" to upload an avatar.

@DeepDiver1975 what do you think ?

**Comment by PVince81** [comment]
Still, the app framework will need to be fixed eventually

**Comment by PVince81** [comment]
Turns out that `$_FILES` isn't set, neither is `$_POST` 
Maybe the JS code is not posting properly.

**Comment by PVince81** [comment]
Aha, let's try this http://stackoverflow.com/questions/12809514/files-array-empty-from-form-submited-in-ie8

**Comment by PVince81** [comment]
Still doesn't work... `$_FILES` is always empty.
Googling...

**Comment by PVince81** [comment]
Ok, turns out it's not a real form post but jquery.fileupload POST from that field. Let's check there.

**Comment by PVince81** [comment]
The "Import root certificate" feature is also broken in IE8 as it uses the same approach. ( @SergioBertolinSG another test case)

**Comment by PVince81** [comment]
From what I see it looks like avatar/certificate upload has NEVER worked in IE8 before, it couldn't, because the "jquery.iframe-transport" wasn't loaded on the page.

But that's not enough, there is another strange issue: because the input element is triggered indirectly through "click", it messes up with the iframe workaround code. According to http://stackoverflow.com/a/14886477 one needs to make the "input type=file" element directly clickable.
In the web UI of the files app, the upload button (the up arrow) is the actual "input type=file" element, this explains why it works in the files app but not in the personal page where the click is indirect...

Now to find a solution to the latter.

**Comment by PVince81** [comment]
So, making the field clickable directly now properly makes the server requests, it uploads the file !

But... the cropper doesn't appear and IE8 wants to download a weird file name. It doesn't make sense. There seems to be another problem.

**Comment by PVince81** [comment]
Okay, it looks like IE8 / the iframe uploading approach needs to return a content-type of "text/plain" even when returning JSON, this was done also for the files app upload: https://github.com/owncloud/core/blob/master/apps/files/ajax/upload.php#L37

Horrorlevel++

**Comment by PVince81** [comment]
Fix will be too intrusive, moving to 8.1.3

**Comment by PVince81** [comment]
PR is here: https://github.com/owncloud/core/pull/18586

You'll notice how many hoops we need to jump through to make this work...

**Comment by cmonteroluque** [comment]
@rperezb can QA test the PR

**Comment by PVince81** [comment]
I'm not sure we really want to backport this, I suggest to keep this in 8.2 only as it's not critical.
Well, it's only bad for people who use IE8 to upload their avatars and security certificates...

**Comment by MorrisJobke** [comment]
> Well, it's only bad for people who use IE8 to upload their avatars and security certificates...

And avatars can also be uploaded to via files app and then be selected.

**Comment by DeepDiver1975** [comment]
this issue it now closed because #18586 is merged - now let's start to discuss the backport.

@cmonteroluque @karlitschek @MTRichards 

I'm with @PVince81 regarding criticality and vote against a backport

**Comment by karlitschek** [comment]
Let's not backport. The workaround should be fine for now. Thanks
